### PR TITLE
Not registering for early error for HTTP/1 upstream connections

### LINKED ISSUE #1679: http/1 flow control issue with no upstream keep alive
There exists a problematic sequence of events:

1. Large response
2. Downstream flow control kicks in
3. Upstream read disable
4. Upstream has actually finished sending response
5. Because we use EPOLLRDHUP we get notified when upstream EOF happens
6. We signal remote close while we still have data when we should probably sit around and wait for flow control to clear.

@alyssawilk it's actually not super clear to me how to fix this. Let's chat Monday. Curious to hear your thoughts on this.

### PR DESCRIPTION
**Pull request by alyssawilk**
This solves the problem where an HTTP upstream connection which is readDisable()d can process the FIN before processing the underlying data

fixes #1679

### CODE REVIEW

File: include/envoy/network/connection.h [hunk 1]
```diff
@@ -123,6 +123,13 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
    */
   virtual void readDisable(bool disable) PURE;
 
+  /* Set if Envoy should detect TCP connection close when readDisable(true) is called.
```
Comment (htuch) [inline include/envoy/network/connection.h]: `/**` to start Doxygen comment on its own line.

File: include/envoy/network/connection.h [hunk 2]
```diff
@@ -123,6 +123,13 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
...
+   * By default, this is true on newly created connections
+   *
+   * @param should_detect supplies if disconnects should be detected when the connection has been
+   * read disabled */
```
Comment (htuch) [inline include/envoy/network/connection.h]: Nit: `*/` on its own line.

File: test/common/http/codec_client_test.cc [hunk 3]
```diff
@@ -167,5 +172,146 @@ TEST_F(CodecClientTest, WatermarkPassthrough) {
...
+  }
+
+protected:
+  Network::MockListenerCallbacks listener_callbacks_;
+  Network::MockConnectionHandler connection_handler_;
+  Network::Address::InstanceConstSharedPtr source_address_;
+  Stats::IsolatedStoreImpl stats_store_;
+  Network::ListenerPtr upstream_listener_;
+  Event::DispatcherPtr dispatcher_;
+  Network::TcpListenSocket socket_{Network::Test::getAnyAddress(Network::Address::IpVersion::v6),
```
Comment (htuch) [inline test/common/http/codec_client_test.cc]: Is this IPv6 only? What about IPv4 only setups? Should the test be paramaterized like we do elsewhere to handle IPv4/IPv6 setup variations?

File: test/common/http/codec_client_test.cc [hunk 4]
```diff
@@ -167,5 +172,146 @@ TEST_F(CodecClientTest, WatermarkPassthrough) {
...
+  close();
+}
+
+// Send a block of data, and then have upstream close the connection.
+// Make sure that the data is passed on as is the network event.
+TEST_F(CodecNetworkTest, SendHeadersAndClose) {
+  createNewStream();
+
+  // Send some header data.
+  std::string full_data = "HTTP/1.1 200 OK\r\ncontent-length: 0\r\n";
```
Comment (htuch) [inline test/common/http/codec_client_test.cc]: Nit: const here and elsewhere.

**Comment by mattklein123** [comment]
LGTM. I think the OSX fail is legit (and similar to other connection test issues we have had) if you could take a quick look. Also, you will need to rewrite your commits with DCO.

**Comment by alyssawilk** [comment]
Probably doing something obviously wrong but any clue why DCO is failing on the last commit?

**Comment by alyssawilk** [comment]
Oh, I think I missed the sign-off on a prior push.   I think I can't fix this without the dreaded rebase.

On the plus side now have the recommended aliases in my .gitconfig so it shouldn't happen again

**Comment by mattklein123** [comment]
@alyssawilk I think you are probably right in your last commit. That's probably an epoll specific feature which OSX/kqueue does not implement. :(

**Pull request merged**